helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java

```diff
@@ -583,6 +587,38 @@ public Set<CapacityNode> getSimpleCapacitySet() {
     return _simpleCapacitySet;
   }
 
+  public void populateSimpleCapacitySetUsage(final Set<String> resourceNameSet,
+      final CurrentStateOutput currentStateOutput) {
+    Map<String, LiveInstance> instanceMap = getLiveInstances();
+    // Convert the assignableNodes to map for quick lookup
+    Map<String, CapacityNode> simpleCapacityMap =
+        _simpleCapacitySet.stream().collect(Collectors.toMap(CapacityNode::getId, node -> node));
```

Comment:
Try to avoid lambda but use for loop. Lambda function is slower and creates a lot of tmp objects, which Helix already experienced before.

Also, will this update the _simpleCapacitySet when you update the map? It is tmp object, right?

Reply:
> Also, will this update the _simpleCapacitySet when you update the map? It is tmp object, right?

It's tmp object acts as dictionary to quickly lookup the `CapacityNode` by name in the set, it won't update `_simpleCapacitySet`